kotlinx-coroutines-core/common/src/flow/StateFlow.kt

```diff
@@ -0,0 +1,410 @@
+    /*
+ * Copyright 2016-2019 JetBrains s.r.o. Use of this source code is governed by the Apache 2.0 license.
+ */
+
+package kotlinx.coroutines.flow
+
+import kotlinx.atomicfu.*
+import kotlinx.coroutines.*
+import kotlinx.coroutines.channels.*
+import kotlinx.coroutines.flow.internal.*
+import kotlinx.coroutines.internal.*
+import kotlin.coroutines.*
+import kotlin.jvm.*
+import kotlin.native.concurrent.*
+
+/**
+ * A [Flow] that represents a read-only state with a single updatable data [value] that emits updates
+ * to the value to its collectors. The current value can be retrieved via [value] property.
+ * The flow of future updates to the value can be observed by collecting values from this flow.
+ *
+ * A [mutable state flow][MutableStateFlow] is created using `StateFlow(value)` constructor function with
+ * the initial value. The value of mutable state flow can be  updated by setting its [value] property.
+ * Updates to the [value] are always [conflated][Flow.conflate]. So a slow collector skips fast updates,
+ * but always collects the most recently emitted value.
+ *
+ * [StateFlow] is useful as a data-model class to represented any kind of state.
+ * Derived values can be defined using various operators on the flows, with [combine] operator being especially
+ * useful to combine values from multiple state flows using arbitrary functions.
+ *
+ * For example, the following class encapsulates an integer state and increments its value on each call to `inc`:
+ *
+ * ```
+ * class CounterModel {
+ *     private val _counter = StateFlow(0) // private mutable state flow
+ *     val counter: StateFlow<Int> get() = _counter // publicly exposed as read-only state flow
+ *
+ *     fun inc() {
+ *         _counter.value++
+ *     }
+ * }
+ * ```
+ *
+ * Having two instances of the above `CounterModel` class one can define the sum of their counters like this:
+ *
+ * ```
+ * val aModel = CounterModel()
+ * val bModel = CounterModel()
+ * val sumFlow: Flow<Int> = aModel.counter.combine(bModel.counter) { a, b -> a + b }
+ * ```
+ *
+ * ### Closing state flow
+ *
+ * A state flow can be optionally [closed][MutableStateFlow.close].
+ * When state flow is closed all its collectors complete normally or with the specified exception.
+ * Closing a state flow transitions it to the terminal state. Once the state flow is closed its value cannot be changed.
+ * The most recent value of the closed state flow is still available via [value] property.
+ * Closing a state is appropriate in situations where the source of the state updates is permanently destroyed.
+ *
+ * **Preview feature**: The design of state flow closing is tentative and could change.
+ *
+ * ### Turning Flow into a StateFlow
+ *
+ * A regular [Flow] is cold. It does not have the concept of the last value and it only becomes active when collected.
+ * It can be made hot by turning it into a [StateFlow] using [stateIn] operator. For example:
+ *
+ * ```
+ * coroutineScope {
```

Comment:
I'm a bit worried about the usability of this one.

When I see state flow, i kind of want to create view models like this:

```
class ViewModel (db : Db) {
    val use = db.userDao().getUser() // returns Flow<User>
                   .stateIn(viewModelScope)
}
```
The problem is that, i cannot do it because `stateIn` is a suspend function which means i cannot assign my values directly like this in the constructor. On the other hand, it has to be a suspend function to inherit the nullability of the type `T` for the `value` field.

I'm not sure what the solution is to be honest. `value` could be always `nullable` (even when `T` is not nullable) but that would make it hard to distinguish whether it has emitted or not when `T` is `nullable`. 

I could also write:
```
class ViewModel(db : Db) {
    val user  = StateFlow<User?>(null)
   viewModelScope.launch {
        db.userDao().getUser().collect { user.value = it)
   }
}
```
but now my UI will always receive a `null` user when observing this even though it is not desired as i'd rather not dispatch a value until it is read.

A third option i can think of is:

```
class ViewModel(db : Db) {
    val user  = viewModelScope.async {
          db.userDao().getUser().stateIn(viewModelScope)
    }
}
```
May not be too bad, the observer just needs to do `user.await().collect`.

I'm actually not sure what the solution is, just wanted to highlight the use case where we want to have a StateFlow that does not have an initial value. Maybe it is not intended in the design and that is totally OK but it would be good to provide a recommended solution on how to handle these.

Any ideas?

Reply:
Idea from [a comment in the related issue](
https://github.com/Kotlin/kotlinx.coroutines/issues/1973#issuecomment-621410369):
Having this function to provide an initial state, that could be `null` or anything.
```kotlin
fun <T> Flow<T>.stateIn(initialValue: T, scope: CoroutineScope): StateFlow<T>
```

Otherwise, your third option seems the best to me. An extra optimization in case the given `StateFlow` is not always needed would be to pass `start = CoroutineStart.LAZY` to `async`.

Reply:
Great idea. I've added `stateIn(scope, value)` that is not suspending. The scope is the first parameter because it reads better this way.

Reply:
We've decided to postpone `stateIn` for a bit and include it into a separate enhancement that would include a whole family of flow-sharing operators as a cohesively designed group.